### DISCUSSION TITLE
즉시 로딩과 지연 로딩이 각각 무엇이고, 유리한 경우에는 어떤 것이 있을까요?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#159] [Spring Data]
.

**Comment by jimi567** [comment]
### 즉시 로딩(Eager)

엔티티가 조회될 때, 연관 관계를 가지는 모든 엔티티를 즉시 로딩하는 방식  
연관 관계에서 fetch = FetchType.EAGER으로 설정하면 즉시 로딩으로 로딩할 수 있습니다.  
실제 DB에 날라가는 쿼리가 복잡해질 수 있고, 한 번의 쿼리로 과도한 데이터를 가져올 수 있게 됩니다.  
연관 관계에서 ~ToOne의 경우 디폴트가 Eager 입니다.  

### 지연 로딩(Lazy)
실제로 필요할 때까지 엔티티를 로딩하지 않는 방식입니다.    
fetch = FetchType.LAZY으로 설정하면 지연 로딩을 할 수 있습니다.  
엔티티를 조회할 때, 연관 관계를 가지는 엔티티는 JPA 프록시 객체로 대체됩니다.  
연관된 엔티티가 실제로 접근될 때 DB에서 조회하게 됩니다.  
연관 관계에서 ~ToMany의 경우 디폴트가 Lazy입니다.  

### 유리한 경우  

1. 즉시 로딩이 유리한 경우  
+ 연관된 엔티티가 자주 함께 사용되는 경우 
+ 데이터의 양이 크지 않아 즉시 로딩을 해도 성능 문제가 미미할 경우  

2. 지연 로딩이 유리한 경우 
+ 연관된 엔티티가 자주 사용되지 않는 경우  
+ 데이터의 양이 많은 경우, 지연 로딩을 통해 성능을 개선할 수 있습니다.  
+ 네트워크 성능을 향상시킬 필요가 있는 경우 네트워크 대역폭을 절약하기 위해 지연 로딩으로 필요한 데이터만 로딩할 수 있습니다.

**Comment by GIVEN53** [comment]
### 즉시 로딩 (Eager Loading)
- 즉시 로딩은 연관된 모든 엔티티도 즉시 로딩하는 방식입니다.
- `@OneToOne`, `@ManyToOne`은 Eager을 기본 전략으로 사용합니다.
- 기본적으로 조인 쿼리를 통해 조회합니다.

#### 장점
필요한 데이터를 한 번에 가져오기 때문에 추가적인 쿼리가 발생하지 않습니다.

> Spring Data JPA로 생성된 동적 쿼리의 경우 추가 쿼리가 발생한다. Spring Data JPA는 eager, lazy 여부와 상관없이 주 엔티티에 대한 동적 쿼리를 생성하기 때문이다.

#### 단점
불필요한 데이터를 모두 로딩하기 때문에 성능 저하가 발생할 수 있습니다.
특히, 연관된 엔티티가 많은 경우 불필요한 데이터 로딩으로 인해 메모리 사용량이 증가하고 응답 속도가 느려질 수 있습니다.

### 지연 로딩 (Lazy Loading)
- 지연 로딩은 연관된 엔티티를 실제로 사용할 때 로딩하는 방식입니다.
- `@OneToMany`, `@ManyToMany`은 Lazy를 기본 전략으로 사용합니다.
- 주 엔티티를 로딩할 때는 연관된 엔티티를 로딩하지 않고, 프록시 객체를 사용합니다.
- 연관된 엔티티는 실제로 필요할 때 추가적인 쿼리를 통해 데이터를 로딩합니다.

#### 장점
필요한 시점에만 데이터를 로딩하기 때문에 초기 로딩 시 불필요한 데이터 로딩을 방지할 수 있습니다.
메모리 사용량이 줄어들고 성능이 향상될 수 있습니다.

#### 단점
연관된 엔티티를 사용할 때마다 추가적인 쿼리가 발생하기 때문에, 연관된 데이터를 많이 사용할 경우 쿼리 성능이 저하될 수 있습니다.
또한, 트랜잭션이 종료된 후에 연관된 엔티티를 로딩하려고 하면 LazyInitializationException이 발생할 수 있습니다.

### 유리한 경우
#### 즉시 로딩이 유리한 경우
- 한 번의 쿼리로 모든 관련 데이터를 가져와야 하는 경우
- 연관된 데이터가 항상 사용되는 경우
- 성능 최적화가 필요하지 않고, 코드가 단순한 것이 더 중요한 경우

#### 지연 로딩이 유리한 경우
- 연관된 데이터가 자주 사용되지 않는 경우
- 초기 로딩 시 성능 최적화가 필요한 경우
- 메모리 사용량을 줄여야 하는 경우

### 추가 키워드
- Spring Data JPA는 동적 쿼리를 어떻게 생성하고 실행하는가
- Reservation - Member가 ManyToOne일 때 로딩 전략에 따라 `findByMember()`, `findByMemberId()`의 쿼리 차이
